walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,151 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  // Return the possible times at which the meeting be set on the day.
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
```

Comment:
I have a general piece of feedback here - and this is in my opinion one of the challenges of using TDD - and that is that the logic of the funciton is getting kind of complex with lots of special cases and edge cases.

Ideally (apart from this exercise) the function would have some helper functions, and read something like the problem specification without needing much explanation, e.g.:

public Collection<TimeRange> query(....) {
    List<Event> requiredEvents = getEventsForAttendees(request.getAttendees());
    List<Event> optionalEvents = getEventsForAttendees(request.getOptionalAttendees());
    Collection<TimeRange> availableTimesForAll = getPossibleTimes(getEventTimeRanges(requiredEvents, optionalEvents));
    if (!availableTimesForAll.empty()) return availableTimesForAll;
    return getPossibleTimes(getEventTimeRanges(requiredEvents));
}

...but I also know that achieving this with the TDD process is kind of hard (which is this exercise), and it tends to encourage lots and lots of edge-case handling.